File: 1474-LongestZigzagPathInABinaryTree/1474-LongestZigzagPathInABinaryTree.cpp
// Last updated: 7/2/2026, 11:16:07 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int longestZigZag(TreeNode* root) {
        int answer = max(helper(root, true, -1), helper(root, false, -1));
        return answer;
    }

    int helper(TreeNode* current, bool isRight, int count){
        if (!current){
            return count;
        }

        if(isRight){
            count++;
            return max(helper(current->right, false, count), helper(current->left, true, 0));
        }

        else{
            count++;
            return max(helper(current->left, true, count), helper(current->right, false, 0));
        }
    }
};

File: 1474-LongestZigzagPathInABinaryTree/1474-LongestZigzagPathInABinaryTree.py
# Last updated: 7/2/2026, 11:16:06 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        return max(self.helper(root, True, -1), self.helper(root, False, -1))

    def helper(self, current, isRight, count):
        if(not current):
            return count;
        if(isRight):
            count += 1
            return max(self.helper(current.right, False, count), self.helper(current.left, True, 0))
        else:
            count += 1
            return max(self.helper(current.left, True, count), self.helper(current.right, False, 0))
        


        